### PR TITLE
[SPARK-46905][SQL] Add dedicated class to keep column definition instead of StructField in Create/ReplaceTable command

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a follow-up of https://github.com/apache/spark/pull/44876 to refactor the code and make it cleaner. The idea is to add a dedicated class for column definition instead of `StructField` in Create/ReplaceTable command. This is more flexible and cleaner than adding an additional default column expression in `CreateTable` command.

### Why are the changes needed?

Code refactor, and makes it easier to fully eliminate the need of a fake analyzer for default value handling in the future. We should make similar code changes to ALTER TABLE command and v1 CREATE TABLE command.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

existing tests

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -3150,61 +3151,44 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
...
-    // Add comment to metadata
-    commentSpec.map(visitCommentSpec).foreach {
-      builder.putString("comment", _)
-    }
-    // Add the 'DEFAULT expression' clause in the column definition, if any, to the column metadata.
-    defaultExpression.map(visitDefaultExpression).foreach { field =>
-      if (conf.getConf(SQLConf.ENABLE_DEFAULT_COLUMNS)) {
-        // Add default to metadata
-        builder.putString(ResolveDefaultColumns.CURRENT_DEFAULT_COLUMN_METADATA_KEY, field)
-        builder.putString(ResolveDefaultColumns.EXISTS_DEFAULT_COLUMN_METADATA_KEY, field)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3190]: The code is moved to `ColumnDefinition#toV1Column`. We only need to encode the default value in `StructField` when running with v1 command.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -3150,61 +3151,44 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
...
-      } else {
-        throw QueryParsingErrors.defaultColumnNotEnabledError(ctx)
-      }
-    }
-    // Add the 'GENERATED ALWAYS AS expression' clause in the column definition, if any, to the
-    // column metadata.
-    generationExpression.map(visitGenerationExpression).foreach { field =>
-      builder.putString(GeneratedColumn.GENERATION_EXPRESSION_METADATA_KEY, field)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3198]: ditto

**Comment by cloud-fan** [comment]
cc @dtenedor @dongjoon-hyun

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (if CI passes).

`ColumnDefinition` looks like a nice contribution to be worthy of a new JIRA, @cloud-fan . If you don't mind, could you make a new JIRA, please?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala [hunk 3]
```diff
@@ -445,6 +444,54 @@ object ResolveDefaultColumns extends QueryErrorsBase with ResolveDefaultColumnsU
...
+      throw QueryCompilationErrors.defaultValuesMayNotContainSubQueryExpressions(
+        statement, colName, default.originalSQL)
+    } else if (default.resolved) {
+      if (!Cast.canUpCast(default.child.dataType, targetType)) {
+        throw QueryCompilationErrors.defaultValuesDataTypeError(
+          statement, colName, default.originalSQL, targetType, default.child.dataType)
+      }
+      // Check passes. We do not check foldable here, as the plan is not optimized yet.
+    } else if (default.references.nonEmpty) {
+      // Ideally we should let the rest of `CheckAnalysis` to report errors about why the default
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:487]:
```suggestion
      // Ideally we should let the rest of `CheckAnalysis` report errors about why the default
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala [hunk 4]
```diff
@@ -445,6 +444,54 @@ object ResolveDefaultColumns extends QueryErrorsBase with ResolveDefaultColumnsU
...
+      targetType: DataType): Unit = {
+    if (default.containsPattern(PLAN_EXPRESSION)) {
+      throw QueryCompilationErrors.defaultValuesMayNotContainSubQueryExpressions(
+        statement, colName, default.originalSQL)
+    } else if (default.resolved) {
+      if (!Cast.canUpCast(default.child.dataType, targetType)) {
+        throw QueryCompilationErrors.defaultValuesDataTypeError(
+          statement, colName, default.originalSQL, targetType, default.child.dataType)
+      }
+      // Check passes. We do not check foldable here, as the plan is not optimized yet.
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:485]:
```suggestion
      // Our analysis check passes here. We do not further inspect whether the
      // expression is `foldable` here, as the plan is not optimized yet.
```

**Review by dtenedor — APPROVED** [review]
LGTM, this makes the parsing and analysis code simpler. It will also let the regular analyzer and optimizer process the default value expressions, which will help Spark users with custom analyzer or optimizer rules as well.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ColumnDefinition.scala [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  def toV1Column: StructField = {
+    val metadataBuilder = new MetadataBuilder().withMetadata(metadata)
+    comment.foreach { c =>
+      metadataBuilder.putString("comment", c)
+    }
+    defaultValue.foreach { default =>
+      // For v1 CREATE TABLE command, we will resolve and execute the default value expression later
+      // in the rule `DataSourceAnalysis`. We just need to put the default value SQL string here.
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ColumnDefinition.scala:67]: Actually I can't recall it now... The existing code did not have comments: https://github.com/apache/spark/pull/44935/files#diff-26fe6b511f4b8aedf5bdffae0d7c03aba69463c09fa4f881d77520125796e0faL3188

[the remark this answers is not in the document]
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ColumnDefinition.scala:67]: Oh this is just moving the code from its previous location :) nevermind then.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ColumnDefinition.scala [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+/**
+ * Column definition for tables. This is an expression so that analyzer can resolve the default
+ * value expression in DDL commands automatically.
+ */
+case class ColumnDefinition(
+    name: String,
+    dataType: DataType,
+    nullable: Boolean = true,
+    comment: Option[String] = None,
+    defaultValue: Option[DefaultValueExpression] = None,
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ColumnDefinition.scala:39]: `QueryPlan.expressions` does support `Option[Expression]`: https://github.com/apache/spark/blob/c468c3d5c685c5a5ecd7caf01f3004addce1f3b6/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala#L276

[the remark this answers is not in the document]
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ColumnDefinition.scala:39]: 👍 missed that, looks like we are OK then.

File: common/utils/src/main/resources/error/error-classes.json [hunk 7]
```diff
@@ -1766,6 +1766,11 @@
           "which requires <expectedType> type, but the statement provided a value of incompatible <actualType> type."
         ]
       },
+      "NOT_CONSTANT" : {
+        "message" : [
+          "which is not a constant expression whose equivalent value is known at query planning time; please update your command to change the column default value to satisfy this constraint and then retry the command again."
```
Comment (yaooqinn) [inline common/utils/src/main/resources/error/error-classes.json:1771]: A bit earful？

**Review by yaooqinn — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Pull request closed without merging**